File: tests/pytest-bdd/steps.py
"""
Step definitions for pytest-bdd.
All steps are pure HTTP-based - no domain knowledge required.
"""
import json
import requests
from pytest_bdd import given, when, then, parsers


# ==================== GIVEN Steps ====================

@given(parsers.parse('the API is available at "{url}"'))
def api_available(context, url):
    """Verify API is running and set base URL."""
    context["base_url"] = url
    resp = requests.get(f"{url}/health")
    assert resp.status_code == 200, f"API not available at {url}"


@given(parsers.parse('a rule template "{name}" exists'))
def template_exists(context, name):
    """Create a rule template with default source."""
    source = 'rule("default").when(f => true).then(f => ({result: "ok"}))'
    _create_template(context, name, source)


@given(parsers.parse('a rule template "{name}" exists with source:'))
def template_exists_with_source(context, name, docstring):
    """Create a rule template with specified source."""
    _create_template(context, name, docstring.strip())


@given(parsers.parse('a policy "{name}" exists using template "{template_name}"'))
def policy_exists(context, name, template_name):
    """Create a policy using specified template."""
    template_id = context["template_ids"].get(template_name)
    assert template_id, f"Template '{template_name}' not found"
    
    payload = {
        "name": name,
        "rule_template_id": template_id,
        "metadata": {}
    }
    resp = requests.post(
        f"{context['base_url']}/api/policies",
        json=payload
    )
    assert resp.status_code == 201, f"Failed to create policy: {resp.text}"
    data = resp.json()
    context["policy_ids"][name] = data["id"]


# ==================== WHEN Steps ====================

@when(parsers.parse('I POST to "{endpoint}" with:'))
def post_to_endpoint(context, endpoint, docstring):
    """Make a POST request with JSON body."""
    payload = json.loads(docstring)
    url = f"{context['base_url']}{endpoint}"
    resp = requests.post(url, json=payload)
    context["last_response"] = resp
    
    # Store IDs for templates/policies
    if resp.status_code in (200, 201):
        data = resp.json()
        if "id" in data:
            if "rule_template_id" in data:
                context["policy_ids"][data.get("name", "")] = data["id"]
            else:
                context["template_ids"][data.get("name", "")] = data["id"]


@when(parsers.parse('I GET "{endpoint}"'))
def get_endpoint(context, endpoint):
    """Make a GET request."""
    url = f"{context['base_url']}{endpoint}"
    resp = requests.get(url)
    context["last_response"] = resp


@when(parsers.parse('I execute policy "{name}" with facts:'))
def execute_policy(context, name, docstring):
    """Execute a policy with given facts."""
    policy_id = context["policy_ids"].get(name)
    assert policy_id, f"Policy '{name}' not found"
    
    facts = json.loads(docstring)
    payload = {
        "policy_id": policy_id,
        "facts": facts
    }
    url = f"{context['base_url']}/api/execute"
    resp = requests.post(url, json=payload)
    context["last_response"] = resp


# ==================== THEN Steps ====================

@then(parsers.parse('the response status should be {status:d}'))
def response_status(context, status):
    """Assert response status code."""
    resp = context["last_response"]
    assert resp.status_code == status, f"Expected {status}, got {resp.status_code}: {resp.text}"


@then(parsers.parse('the response should contain "{text}"'))
def response_contains(context, text):
    """Assert response body contains text."""
    resp = context["last_response"]
    assert text in resp.text, f"Response does not contain '{text}': {resp.text}"


@then(parsers.parse('the response field "{field}" should be {value:d}'))
def response_field_int(context, field, value):
    """Assert integer field value."""
    data = context["last_response"].json()
    assert data.get(field) == value, f"Field '{field}' is {data.get(field)}, expected {value}"


@then(parsers.parse('the response field "{field}" should be null'))
def response_field_null(context, field):
    """Assert field is null."""
    data = context["last_response"].json()
    assert data.get(field) is None, f"Field '{field}' is not null: {data.get(field)}"


@then("the response should be a list")
def response_is_list(context):
    """Assert response is a list."""
    data = context["last_response"].json()
    assert isinstance(data, list), f"Response is not a list: {type(data)}"


@then("the execution should succeed")
def execution_succeeds(context):
    """Assert execution succeeded."""
    data = context["last_response"].json()
    assert data.get("success") is True, f"Execution failed: {data}"


@then("the condition should be met")
def condition_met(context):
    """Assert condition was met."""
    data = context["last_response"].json()
    assert data.get("condition_met") is True, f"Condition not met: {data}"


@then("the condition should NOT be met")
def condition_not_met(context):
    """Assert condition was not met."""
    data = context["last_response"].json()
    assert data.get("condition_met") is False, f"Condition unexpectedly met: {data}"


@then(parsers.parse('the output field "{field}" should be {value:d}'))
def output_field_value(context, field, value):
    """Assert output_facts field value."""
    data = context["last_response"].json()
    output = data.get("output_facts", {})
    assert output.get(field) == value, f"Output field '{field}' is {output.get(field)}, expected {value}"


# ==================== Helpers ====================

def _create_template(context, name, source):
    """Helper to create a rule template."""
    payload = {"name": name, "source": source}
    resp = requests.post(
        f"{context['base_url']}/api/rule-templates",
        json=payload
    )
    assert resp.status_code == 201, f"Failed to create template: {resp.text}"
    data = resp.json()
    context["template_ids"][name] = data["id"]

File: tests/rstest-bdd/steps.rs
//! rstest-bdd Step Definitions for Policy Hub API
//!
//! Demonstrates `rstest-bdd` v0.5 with Gherkin `.feature` files,
//! `#[given]`/`#[when]`/`#[then]` macros, and `Slot`-based state.
//!
//! Runs via `cargo test --test rstest_bdd_steps -p policy-hub-api`

use reqwest::blocking::Client;
use rstest_bdd::state::{ScenarioState, Slot};
use rstest_bdd_macros::{given, when, then, scenario, ScenarioState};
use serde_json::{json, Value};
use std::collections::HashMap;
use std::time::Duration;

// ==================== State ====================

#[derive(Default, ScenarioState)]
struct PolicyState {
    client: Slot<Client>,
    base_url: Slot<String>,
    template_ids: Slot<HashMap<String, String>>,
    policy_ids: Slot<HashMap<String, String>>,
    last_response: Slot<Value>,
    last_status: Slot<u16>,
}

impl PolicyState {
    fn ensure_init(&self) {
        if self.client.is_empty() {
            self.client.set(
                Client::builder()
                    .timeout(Duration::from_secs(10))
                    .build()
                    .expect("Failed to create HTTP client"),
            );
            self.base_url.set(
                std::env::var("API_BASE_URL")
                    .unwrap_or_else(|_| "http://localhost:8080".to_string()),
            );
            self.template_ids.set(HashMap::new());
            self.policy_ids.set(HashMap::new());
        }
    }

    fn base(&self) -> String {
        self.base_url
            .with_ref(|u| u.clone())
            .unwrap_or_else(|| "http://localhost:8080".to_string())
    }
}

// ==================== GIVEN Steps ====================

#[given("the API server is running")]
fn the_api_server_is_running(state: &PolicyState) {
    state.ensure_init();
    let base = state.base();
    state.client.with_ref(|c| {
        let resp = c
            .get(format!("{}/health", base))
            .send()
            .expect("API server is not running");
        assert!(resp.status().is_success(), "Health check failed");
    });
}

#[given("a rule template {name:string} exists with source:")]
fn a_rule_template_exists_with_source(state: &PolicyState, name: String, docstring: String) {
    state.ensure_init();
    let base = state.base();
    let (status, id) = state.client.with_ref(|c| {
        let resp = c
            .post(format!("{}/api/rule-templates", base))
            .json(&json!({
                "name": &name,
                "source": &docstring
            }))
            .send()
            .expect("Failed to create template");

        let status = resp.status().as_u16();
        let body: Value = resp.json().expect("Failed to parse response");
        let id = body["id"].as_str().expect("No id in response").to_string();
        (status, id)
    }).unwrap();

    assert_eq!(status, 201, "Template creation failed");
    state.template_ids.with_mut(|map| {
        map.insert(name, id);
    });
}

#[given("a policy {name:string} exists using template {template:string}")]
fn a_policy_exists_using_template(state: &PolicyState, name: String, template: String) {
    state.ensure_init();
    let base = state.base();
    let template_id = state.template_ids.with_ref(|map| {
        map.get(&template)
            .cloned()
            .unwrap_or_else(|| panic!("Template '{}' not found", template))
    }).unwrap();

    let (status, id) = state.client.with_ref(|c| {
        let resp = c
            .post(format!("{}/api/policies", base))
            .json(&json!({
                "name": &name,
                "rule_template_id": &template_id,
                "metadata": {}
            }))
            .send()
            .expect("Failed to create policy");

        let status = resp.status().as_u16();
        let body: Value = resp.json().expect("Failed to parse response");
        let id = body["id"].as_str().expect("No id in response").to_string();
        (status, id)
    }).unwrap();

    assert_eq!(status, 201, "Policy creation failed");
    state.policy_ids.with_mut(|map| {
        map.insert(name, id);
    });
}

// ==================== WHEN Steps ====================

#[when("I execute policy {name:string} with amount {amount:int}")]
fn i_execute_policy_with_amount(state: &PolicyState, name: String, amount: i64) {
    let base = state.base();
    let policy_id = state.policy_ids.with_ref(|map| {
        map.get(&name)
            .cloned()
            .unwrap_or_else(|| panic!("Policy '{}' not found", name))
    }).unwrap();

    let (status, body) = state.client.with_ref(|c| {
        let resp = c
            .post(format!("{}/api/execute", base))
            .json(&json!({
                "policy_id": &policy_id,
                "facts": { "amount": amount }
            }))
            .send()
            .expect("Failed to execute policy");

        let status = resp.status().as_u16();
        let body: Value = resp.json().expect("Failed to parse response");
        (status, body)
    }).unwrap();

    state.last_status.set(status);
    state.last_response.set(body);
}

// ==================== THEN Steps ====================

#[then("the execution should succeed")]
fn the_execution_should_succeed(state: &PolicyState) {
    state.last_response.with_ref(|resp| {
        assert_eq!(resp["success"], true, "Execution did not succeed: {:?}", resp);
    }).expect("No response");
}

#[then("the condition should be met")]
fn the_condition_should_be_met(state: &PolicyState) {
    state.last_response.with_ref(|resp| {
        assert_eq!(resp["condition_met"], true, "Condition was not met: {:?}", resp);
    }).expect("No response");
}

#[then("the condition should NOT be met")]
fn the_condition_should_not_be_met(state: &PolicyState) {
    state.last_response.with_ref(|resp| {
        assert_eq!(resp["condition_met"], false, "Condition was unexpectedly met: {:?}", resp);
    }).expect("No response");
}

#[then("the discount should be {expected:int}")]
fn the_discount_should_be(state: &PolicyState, expected: i64) {
    state.last_response.with_ref(|resp| {
        let discount = resp["output_facts"]["discount"]
            .as_i64()
            .expect("No discount in output_facts");
        assert_eq!(discount, expected, "Unexpected discount value");
    }).expect("No response");
}

// ==================== Fixture ====================

#[rstest::fixture]
fn state() -> PolicyState {
    let s = PolicyState::default();
    s.ensure_init();
    s
}

// ==================== Scenario Bindings ====================

#[scenario(
    "tests/rstest-bdd-features/execution.feature",
    name = "Execute policy when condition is met"
)]
fn execute_condition_met(state: PolicyState) {}

#[scenario(
    "tests/rstest-bdd-features/execution.feature",
    name = "Execute policy when condition is NOT met"
)]
fn execute_condition_not_met(state: PolicyState) {}

#[scenario(
    "tests/rstest-bdd-features/execution.feature",
    name = "Execute policy at boundary value"
)]
fn execute_boundary_value(state: PolicyState) {}
